server/src/models/timestamps.rs

```diff
@@ -0,0 +1,352 @@
+            leading_one_bits += 1;
+        }
+
+        // Any leftover bits in residual_timestamps are set to one. Thus, a
+        // sequence of one bits followed by fewer bits than specified by the
+        // flag means that all residual timestamps have been decompressed.
+        if leading_one_bits != 0 && bits.remaining_bits() < 7 {
+            break;
+        }
+
+        let delta = match leading_one_bits {
+            0 => last_delta,                                               // Flag is 0.
+            1 => read_decode_and_compute_delta(&mut bits, 7, last_delta),  // Flag is 10.
+            2 => read_decode_and_compute_delta(&mut bits, 9, last_delta),  // Flag is 110.
+            3 => read_decode_and_compute_delta(&mut bits, 12, last_delta), // Flag is 1110.
+            4 => last_delta + bits.read_bits(32),                          // Flag is 1111.
+            _ => panic!("Unknown encoding of timestamps."),
+        };
+
+        timestamp += delta as i64;
+        timestamp_builder.append_value(timestamp);
+        last_delta = delta;
+    }
+
+    // Add the last timestamp stored as `end_time` in the segment.
+    timestamp_builder.append_value(end_time);
+    timestamp_builder.finish()
+}
+
+/// Read the next delta-of-delta as `bits_to_read` from `bits`, decode the
+/// delta-of-delta, and add it to `last_delta` to compute the next delta.
+/// [`read_decode_and_compute_delta`] is implemented based on [code published by
+/// Jerome Froelich] under MIT.
+///
+/// [code published by Jerome Froelich]: code published by Jerome Froelich
+fn read_decode_and_compute_delta(bits: &mut BitReader, bits_to_read: u8, last_delta: u32) -> u32 {
+    let encoded_delta_of_delta = bits.read_bits(bits_to_read);
+    let delta_of_delta = if encoded_delta_of_delta > (1 << (bits_to_read - 1)) {
+        encoded_delta_of_delta | (u32::max_value() << bits_to_read)
+    } else {
+        encoded_delta_of_delta
+    };
+    // Wrapping_add() ensure negative values are handled correctly
+    last_delta.wrapping_add(delta_of_delta)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Tests for compress_residual_timestamps() and decompress_all_timestamps().
+    #[test]
+    fn compress_timestamps_for_time_series_with_zero_one_or_two_timestamps() {
+        let mut uncompressed_timestamps_builder = TimestampBuilder::new(3);
+
+        uncompressed_timestamps_builder.append_slice(&[]);
+        assert!(compress_residual_timestamps(&uncompressed_timestamps_builder.finish()).is_empty());
+
+        uncompressed_timestamps_builder.append_slice(&[100]);
+        assert!(compress_residual_timestamps(&uncompressed_timestamps_builder.finish()).is_empty());
+
+        uncompressed_timestamps_builder.append_slice(&[100, 300]);
+        assert!(compress_residual_timestamps(&uncompressed_timestamps_builder.finish()).is_empty());
+    }
+
+    #[test]
+    fn compress_and_decompress_timestamps_for_a_regular_time_series() {
```

Comment:
Since this test and the next only differs in the initialization of `uncompressed_timestamps`, maybe create a utility function to avoid duplicated code.

Reply:
Fixed.